### PR TITLE
feat(aclass): propose ABAP OO parser package (Chevrotain, Ch)

### PR DESCRIPTION
**Pull request by ThePlenkov**
## Summary

Introduces **`@abapify/aclass`** — a lexer + statement parser + typed AST for ABAP OO (`.clas.abap` / `.intf.abap`) — and wires it into `@abapify/openai-codegen` as a CI parse-gate. Symmetric to `@abapify/acds` (CDS parser) and the reverse direction of `@abapify/abap-ast` (ABAP emitter).

Live verification on TRL (BTP Steampunk): deploy → activation pre-audit clean → `adt aunit -c ZCL_PS3_CLIENT_TESTS` 3/3 pass.

## What landed (Waves 0–3)

| Wave | Scope | File set |
|------|-------|----------|
| **0** | Package skeleton + Chevrotain lexer | `packages/aclass/{package.json, tsconfig, tsdown.config, vitest.config, eslint.config}`, `src/tokens.ts`, `src/lex.ts`, `src/errors.ts`, `tests/lex.test.ts` |
| **1** | Parser + typed AST | `src/ast.ts`, `src/parser.ts` (statement splitter — not a Chevrotain CstParser — see AGENTS.md for rationale), `tests/parse-interface.test.ts` |
| **2** | Petstore3 fixture corpus + structural roundtrip | `tests/fixtures.test.ts`, `tests/roundtrip.test.ts` |
| **3** | Consumer wiring | `src/assert.ts` (`assertCleanParse`, `AclassParseError`), `packages/openai-codegen/tests/aclass-parse-gate.test.ts` |

## Scope

**In** (structural):
- Class / interface DEFINITION + IMPLEMENTATION headers
- Sections (PUBLIC / PROTECTED / PRIVATE)
- Member declarations: METHODS, DATA, CLASS-DATA, TYPES (incl. STANDARD/SORTED/HASHED TABLE, BEGIN/END OF), CONSTANTS, EVENTS, ALIASES, INTERFACES
- Inheritance (INHERITING FROM), create visibility, FOR TESTING
- Method IMPLEMENTATIONS with opaque body slice (start/end offsets + raw text); body statements are NOT parsed
- ABAPDoc lines attached to the following declaration
- **Keywords-as-names** — ABAP reuses reserved words as identifiers in declaration positions (`DATA data TYPE i.`, `REF TO data`, struct fields named `type`/`data`, etc.); the parser accepts any identifier-shaped token image.

**Out** (explicitly deferred): method body statement parsing, top-level DATA/TYPES outside classes, FORM/PERFORM, macros, includes, CDS annotations inside classes, SELECT/UPDATE/INSERT, walker helpers, semantic validators.

## Tests

- **`aclass`**: 48/48 (lex smoke, per-topic grammar, keywords-as-names, petstore3 fixtures, structural roundtrip, body-span accuracy, `assertCleanParse` helper)
- **`openai-codegen`**: 153/153 (includes new `aclass-parse-gate.test.ts` — 11 assertions — every generated `.clas.abap`/`.intf.abap` must parse cleanly)
- **`abap-ast`**: 115/115 (no regressions)

## Live verification

```bash
bunx adt deploy --source samples/petstore3-client/generated/abapgit --package ZPEPL --activate --unlock
# → ✅ 4 objects activated

bunx adt fetch -X POST ... '/sap/bc/adt/activation?method=activate&preauditRequested=false'
# → activationExecuted="false" generationExecuted="true" (no errors)

bunx adt aunit -c ZCL_PS3_CLIENT_TESTS --format json
# → totalTests: 3, passCount: 3, failCount: 0, errorCount: 0
```

## Out of scope for this PR (tracked in `openspec/changes/add-aclass-parser/tasks.md` / spec)

- Full byte-for-byte `print(parse(src)) === src` roundtrip (current test is structural)
- AST walker helpers (`walkDefinitions`, `walkMembers`, …)
- Converters `aclass → abap-ast` / `abap-ast → aclass`
- `CLASS-EVENTS` parameter types, `ALIASES … IMPLEMENTED BY`

## Risks / open questions

- **Parser uses a statement splitter, not a Chevrotain CstParser.** This is a deliberate choice documented in `packages/aclass/AGENTS.md`; the wave plan in `tasks.md` originally called for a CstParser, but the statement-splitter approach made the parser half the size with no loss of capability for the structural subset. If a future need for byte-exact recovery / partial CST inspection emerges, the CstParser can be slotted in behind the same `parse()` entry point.
- **Fixture corpus is the live petstore3 generator output**, not a frozen snapshot. Change either side of that contract and the other side's tests flip — intended coupling, documented in `packages/aclass/AGENTS.md`.

Generated with [Devin](https://cli.devin.ai/docs)

## Release Notes

* **New Features**
  * Introduced `@abapify/aclass`, a new parser for ABAP class and interface definitions. Parses `.clas.abap` and `.intf.abap` source files into a typed Abstract Syntax Tree, with method bodies preserved as opaque source slices. Includes standardized error reporting and clean-parse validation for integration workflows.

### CODE REVIEW

File: openspec/changes/add-aclass-parser/proposal.md [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+   `FINAL`, `ABSTRACT`, `CREATE`, `IMPORTING`, `EXPORTING`, `CHANGING`,
+   `RETURNING`, `RAISING`, `TYPE`, `REF`, `TO`, `REDEFINITION`,
+   `OPTIONAL`, `DEFAULT`, `VALUE`, `BEGIN`, `END`, `OF`, `STANDARD`,
+   `SORTED`, `HASHED`, `TABLE`, `STRUCTURE`, `WITH`, `KEY`, `EMPTY`,
+   `UNIQUE`, `NON-UNIQUE`), identifiers, string literals, integer
+   literals, `ABAPDocLine` (`"!`), line comments (`"`), star comments
+   (`*`), end-of-statement dot, comma, colon, paren open/close,
+   backtick literal, hyphen, whitespace.
+3. **Parser** — a Chevrotain `CstParser` with production rules limited
+   to: `cdsClassDef`, `cdsClassImplementation`, `cdsInterfaceDef`, each
```
[the remark this answers is not in the document]
Comment (ThePlenkov) [reply openspec/changes/add-aclass-parser/proposal.md]: Fixed in e1b2047d — proposal.md now uses `classDef` / `classImpl` / `interfaceDef` and the surrounding paragraph describes the shipped statement-splitter approach instead of the Chevrotain CstParser form the original proposal sketched.

File: openspec/changes/add-aclass-parser/specs/aclass/spec.md [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+1. **Scope is strictly structural.** The parser recognises class /
+   interface _declarations_: headers, sections, member declarations
+   (methods, attributes, events, types, constants, aliases),
+   inheritance, implements lists, and the opaque bodies of method
+   implementations. It does NOT parse statements inside a method
+   body, nor does it parse top-level DATA/TYPES outside classes.
+2. **Method body is an opaque source slice.** Every `MethodImpl`
+   node carries the raw text between its opening `METHOD <name>.`
+   and its terminating `ENDMETHOD.`, plus a `{ startOffset, endOffset,
+startLine, startColumn }` span. Callers that need structured
```
[the remark this answers is not in the document]
Comment (ThePlenkov) [reply openspec/changes/add-aclass-parser/specs/aclass/spec.md:17]: Fixed in e1b2047d — joined the run-on line and moved `errors: ParseError[]` into a dedicated `ParseResult` block so it no longer appears to live inside `AbapSourceFile`.

File: packages/openai-codegen/tests/aclass-parse-gate.test.ts [hunk 3]
```diff
@@ -0,0 +1,98 @@
+/**
+ * CI gate: every generated `.clas.abap` / `.intf.abap` that the
+ * openai-codegen pipeline ships in `samples/petstore3-client` MUST
+ * round-trip through `@abapify/aclass` cleanly.
+ *
+ * This asserts that our emitter output stays *inside* the structural
+ * subset of ABAP that `aclass` understands. If someone changes the
+ * emitter to produce a shape the parser doesn't recognise, this test
+ * flips red — the fix is either to extend the parser's grammar or to
+ * revisit the emitter output.
```
[the remark this answers is not in the document]
Comment (ThePlenkov) [reply packages/openai-codegen/tests/aclass-parse-gate.test.ts:10]: Fixed in e1b2047d — rewrote the PR body to describe what actually landed (Waves 0–3 + live TRL verification). The earlier "docs-only" framing was correct for the first commit but stopped being true at `fc76f504`.

File: packages/aclass/src/parser.ts [hunk 4]
```diff
@@ -0,0 +1,1285 @@
...
+      return {
+        kind: 'MethodImpl',
+        name: nameTok.image,
+        body,
+        bodySpan: {
+          startOffset: bodyStart,
+          endOffset: bodyEnd,
+          startLine: methodTok.startLine ?? 1,
+          startColumn: methodTok.startColumn ?? 1,
+        },
```
[the remark this answers is not in the document]
Comment (ThePlenkov) [reply packages/aclass/src/parser.ts:468]: Fixed in e1b2047d — `MethodImpl.bodySpan.startLine/startColumn` now resolve the first token at/after `bodyStart` via a new `Cursor.firstTokenAtOrAfter()` helper. Added tests for a normal body (line points at content, not at `METHOD`) and for the empty-body edge case.

File: packages/aclass/src/parser.ts [hunk 5]
```diff
@@ -0,0 +1,1285 @@
...
+  'abap_false',
+  'sy-subrc',
+  'sy-tabix',
+]);
+function isBuiltinName(name: string): boolean {
+  return BUILTIN_ABAP_TYPES.has(name.toLowerCase());
+}
+
+// Silence the unused-name warning on the legacy import.
+void T;
```
[the remark this answers is not in the document]
Comment (ThePlenkov) [reply packages/aclass/src/parser.ts]: Fixed in e1b2047d — dropped both the unused `import * as T from './tokens'` and the `void T;` silencer. No remaining noise hiding genuine unused-import signals.

File: openspec/changes/add-aclass-parser/tasks.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
+# Tasks — `@abapify/aclass`
+
+## Wave 0 — skeleton + lexer (first PR)
+
+1. [ ] Scaffold `packages/aclass/` (package.json, tsconfig, tsdown.config.ts, vitest.config.ts, eslint.config.js, project.json for Nx) mirroring `packages/acds/`.
+2. [ ] Wire `@abapify/aclass` into `tsconfig.base.json` paths and the root workspace `package.json` workspaces array.
+3. [ ] Write `src/tokens.ts` — Chevrotain token definitions for the ABAP OO subset listed in the spec. Keyword order must put longer keywords before their prefixes (`INTERFACE` before `INTERFACES`, `CLASS-DATA` before `CLASS`, `NON-UNIQUE` before `UNIQUE`) so the lexer doesn't split them.
```
[the remark this answers is not in the document]
Comment (ThePlenkov) [reply openspec/changes/add-aclass-parser/tasks.md]: Fixed in e1b2047d — the note now reads `INTERFACES` before `INTERFACE` (matching the lexer source and `packages/aclass/AGENTS.md`). Also flipped every `[ ]` to `[x]` since all waves landed in this PR.

File: openspec/changes/add-aclass-parser/specs/aclass/spec.md [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+  form.
+
+## AST shape
+
+```
+AbapSourceFile {
+  kind: 'AbapSourceFile',
+  source: string,                      // original input, for offset math
+  definitions: AbapDefinition[],
+  errors: ParseError[],
```
[the remark this answers is not in the document]
Comment (ThePlenkov) [reply openspec/changes/add-aclass-parser/specs/aclass/spec.md]: Fixed in e1b2047d — moved `errors: ParseError[]` out of `AbapSourceFile` and into a new `ParseResult` block. AST shape and impl are now aligned.

File: packages/aclass/AGENTS.md [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+| File             | Purpose                                                                 |
+| ---------------- | ----------------------------------------------------------------------- |
+| `src/index.ts`   | Public exports. Re-exports `tokenize`, lexer, token types, error shape. |
+| `src/tokens.ts`  | Chevrotain token definitions (keywords, symbols, literals, comments).   |
+| `src/lex.ts`     | `tokenize(source)` — thin wrapper that normalises lex errors.           |
+| `src/errors.ts`  | Chevrotain lex + parse errors → stable `ParseError` shape.              |
+| `src/parser.ts`  | [Wave 1] CstParser rules.                                               |
+| `src/visitor.ts` | [Wave 1] CST → typed AST.                                               |
+| `src/ast.ts`     | [Wave 1] AST node interfaces.                                           |
```
[the remark this answers is not in the document]
Comment (ThePlenkov) [reply packages/aclass/AGENTS.md]: Fixed in e1b2047d — rewrote the "Current status" section to describe all four shipped waves, updated the architecture diagram to show the statement-splitter pipeline (no phantom Visitor / CstParser stages), and refreshed the key-files table to list the actual source files (`src/parser.ts`, `src/assert.ts`).

File: packages/aclass/AGENTS.md [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+
+## Anti-patterns to avoid
+
+| Don't                                    | Do instead                                                  |
+| ---------------------------------------- | ----------------------------------------------------------- |
+| Hand-rolled string scanning              | Extend `allTokens`                                          |
+| `new RegExp(...)` for each keyword match | `createToken({ pattern: /word/i, longer_alt: Identifier })` |
+| Add `then` property on AST nodes         | Use `thenBody` (JS thenable clash — see `abap-ast`)         |
+| Depend on `@abapify/abap-ast` at runtime | Devdep only; aclass owns its AST                            |
+| `throw` on malformed input               | Return `{ ast, errors }`                                    |
```
[the remark this answers is not in the document]
Comment (ThePlenkov) [reply packages/aclass/AGENTS.md]: Fixed in e1b2047d — removed the `new RegExp(...)` anti-pattern row. The `kw()` helper legitimately constructs dynamic regexes with a `REGEX_META` escape table (CodeQL `js/incomplete-sanitization` clean), and the old guidance was contradicting what the shipped code does.

File: packages/aclass/src/parser.ts [hunk 10]
```diff
@@ -0,0 +1,1285 @@
...
+  const startTok = toks[i];
+  if (
+    startTok.tokenType.name !== 'Identifier' &&
+    startTok.tokenType.name !== 'Value'
+  ) {
+    // `Value` shouldn't appear here, but harmless.
+    // Keep collecting identifiers — but at least the head must be an identifier.
+    c.report(`expected type name, got "${startTok.image}"`, startTok);
+    return null;
+  }
```
[the remark this answers is not in the document]
Comment (ThePlenkov) [reply packages/aclass/src/parser.ts]: Fixed in e1b2047d — `consumeTypeRef` now uses a new `isNameLike()` type-guard that accepts any identifier-shaped token image, including keywords the grammar re-purposes in declaration positions. `REF TO data`, `TYPE type`, and qualified refs like `zif_y=>data` now parse correctly. Tests in `tests/parse-interface.test.ts` cover the head, the `REF TO` path, and the qualified-tail path. Same helper is applied in `parseFieldRun` and `consumeMethodParam`.

File: packages/aclass/src/parser.ts [hunk 11]
```diff
@@ -0,0 +1,1285 @@
...
+    if (toks[i]?.tokenType.name === 'Comma') {
+      i += 1;
+      continue;
+    }
+    const nameTok = toks[i];
+    if (!nameTok || nameTok.tokenType.name !== 'Identifier') {
+      // Unexpected — skip to next comma to recover
+      while (i < toks.length && toks[i].tokenType.name !== 'Comma') i += 1;
+      continue;
+    }
```
[the remark this answers is not in the document]
Comment (ThePlenkov) [reply packages/aclass/src/parser.ts:1037]: Fixed in e1b2047d — `parseFieldRun` now uses `isNameLike()` so struct fields named `type`, `data`, `value`, etc. parse correctly. Test covers `TYPES: BEGIN OF row, type TYPE string, data TYPE i, value TYPE string, END OF row.`

File: packages/aclass/src/ast.ts [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+}
+
+export type ClassMember =
+  | MethodDecl
+  | AttributeDecl
+  | TypeDecl
+  | ConstantDecl
+  | InterfaceStmt
+  | AliasDecl
+  | RawMember;
```
[the remark this answers is not in the document]
Comment (ThePlenkov) [reply packages/aclass/src/ast.ts:103]: Fixed in e200ccd9 — added `EventDecl` to the `ClassMember` union and the `ast.ts` interface, plus a `parseEventDecl()` that handles both `EVENTS` and `CLASS-EVENTS` with optional `EXPORTING VALUE(x) TYPE t` parameters. New test in `tests/parse-interface.test.ts` parses `EVENTS changed EXPORTING VALUE(payload) TYPE string.` and asserts `name`, `isClassEvent`, and the `exporting` param shape.

File: packages/aclass/tests/fixtures.test.ts [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+  it.each(files)(
+    'parses $name with no lex errors and at least one top-level definition',
+    ({ source, name }) => {
+      const { ast, errors } = parse(source);
+      // Lex errors would indicate an unknown character in real output —
+      // every real token must be in the lexer's vocabulary.
+      const lexErrors = errors.filter((e) =>
+        /Unexpected character/.test(e.message),
+      );
+      expect(lexErrors, `lex errors in ${name}`).toEqual([]);
```
[the remark this answers is not in the document]
Comment (ThePlenkov) [reply packages/aclass/tests/fixtures.test.ts]: Fixed in e200ccd9 — `fixtures.test.ts` now asserts `expect(errors).toEqual([])` for every fixture, not just the lex subset. Stricter gate immediately caught a latent bug: the parser was silently accepting unterminated classes; now it emits an "unterminated CLASS …" diagnostic. Also exposed that `locals_def.abap` uses `CLASS … DEFINITION DEFERRED.` forward declarations and method names like `to` — both are now handled (see follow-up fixes in the same commit).

File: packages/aclass/tests/parse-interface.test.ts [hunk 14]
```diff
@@ -0,0 +1,348 @@
...
+    ].join('\n');
+    const { ast, errors } = parse(src);
+    expect(errors).toEqual([]);
+    const cls = ast.definitions[0];
+    if (cls.kind !== 'ClassDef') return;
+    const raw = cls.sections[0].members[0];
+    expect(raw.kind).toBe('RawMember');
+    if (raw.kind !== 'RawMember') return;
+    expect(raw.source).toContain('EVENTS');
+    expect(raw.source).toContain('changed');
```
[the remark this answers is not in the document]
Comment (ThePlenkov) [reply packages/aclass/tests/parse-interface.test.ts]: Fixed in e200ccd9 — replaced the EVENTS-as-RawMember test with two new tests: a positive `EventDecl` assertion, and a genuinely-unknown (`WILDCARD …`) `RawMember` test that still validates the fallback path. `EVENTS` is no longer locked in as "unknown".

File: packages/aclass/tests/parse-interface.test.ts [hunk 15]
```diff
@@ -0,0 +1,348 @@
...
+  it('is robust against a missing ENDCLASS', () => {
+    const { ast, errors } = parse(
+      'CLASS zcl_x DEFINITION PUBLIC.\n  PUBLIC SECTION.\n',
+    );
+    // Should not throw; produces at least the opening ClassDef.
+    expect(ast.definitions.length).toBeGreaterThan(0);
+    // Errors may or may not include diagnostics — the contract is only
+    // "doesn't throw and produces a best-effort AST".
+    void errors;
+  });
```
[the remark this answers is not in the document]
Comment (ThePlenkov) [reply packages/aclass/tests/parse-interface.test.ts]: Fixed in e200ccd9 — `parseClassDef` / `parseClassImpl` now emit an "unterminated CLASS …" diagnostic when `ENDCLASS.` never arrives. The test requires `errors.length > 0` with `severity === 'error'`, so silent information loss is no longer possible.

File: packages/openai-codegen/tests/aclass-parse-gate.test.ts [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+    .filter((f) => f.endsWith('.abap'))
+    .map((f) => ({ f, src: readFileSync(join(petstore3Dir, f), 'utf8') }));
+
+  it.each(files)('$f parses without lex errors', ({ f, src }) => {
+    const { errors } = parse(src);
+    const lexErrors = errors.filter((e) =>
+      /Unexpected character/i.test(e.message),
+    );
+    expect(lexErrors, `unexpected characters in ${f}`).toEqual([]);
+  });
```
[the remark this answers is not in the document]
Comment (ThePlenkov) [reply packages/openai-codegen/tests/aclass-parse-gate.test.ts]: Fixed in e200ccd9 — `openai-codegen/tests/aclass-parse-gate.test.ts` now runs every generator output through BOTH `assertCleanParse` (aclass) AND `@abaplint/core.Registry` (second opinion). Gated on `parser_error` keys only, matching the convention of the neighbouring `emit-exception-class.test.ts` / `emit-implementation-class.test.ts`. 6 new assertions; all pass.

File: packages/aclass/src/parser.ts [hunk 17]
```diff
@@ -0,0 +1,1357 @@
...
+
+function parseMethodImpl(c: Cursor): MethodImpl | null {
+  const methodTok = c.current();
+  const header = c.collectStatement();
+  // header.tokens: [Method, Ident, …]
+  const nameTok = header.tokens[1];
+  if (!nameTok || nameTok.tokenType.name !== 'Identifier') {
+    c.report('expected method name after METHOD', nameTok ?? methodTok);
+    return null;
+  }
```
[the remark this answers is not in the document]
Comment (ThePlenkov) [reply packages/aclass/src/parser.ts]: Fixed in cfe5839d — `parseMethodImpl` uses a new `readQualifiedName(toks, start)` helper that accepts keyword-as-name tokens via `isNameLike()`. `METHOD to.` (and any other keyword-named method) now parses into a proper `MethodImpl`. New test covers `METHOD zif_petstore3~add_pet.`; same helper is applied to `parseMethodDecl`.

File: packages/aclass/src/parser.ts [hunk 18]
```diff
@@ -0,0 +1,1357 @@
...
+  const head = c.current();
+  const isClassMethod = head.tokenType.name === 'ClassMethods';
+  const stmt = c.collectStatement();
+  const toks = stmt.tokens;
+  // toks: [Methods|ClassMethods, Ident, …]
+  const nameTok = toks[1];
+  if (!nameTok || nameTok.tokenType.name !== 'Identifier') {
+    c.report('expected method name', nameTok ?? head);
+    return null;
+  }
```
[the remark this answers is not in the document]
Comment (ThePlenkov) [reply packages/aclass/src/parser.ts]: Fixed in cfe5839d — `parseMethodDecl` now routes through the same `readQualifiedName()` + `isNameLike()` helpers used by `parseMethodImpl`. `METHODS to IMPORTING target TYPE REF TO data.` (from `zcl_petstore3.clas.locals_def.abap:67`) now parses into a proper `MethodDecl`, and `METHODS zif_foo~bar REDEFINITION.` preserves the full qualified name plus `isRedefinition: true`. Test added.

File: packages/aclass/src/parser.ts [hunk 19]
```diff
@@ -0,0 +1,1357 @@
...
+      const body = c['sliceSource'](bodyStart, bodyEnd);
+      // Locate the first non-whitespace line/column inside the body so
+      // bodySpan points at the actual content, not at the `METHOD`
+      // keyword that terminated with the `.` before bodyStart.
+      const firstBodyTok = c.firstTokenAtOrAfter(bodyStart);
+      const bodyLine = firstBodyTok?.startLine ?? methodTok.startLine ?? 1;
+      const bodyCol = firstBodyTok?.startColumn ?? 1;
+      return {
+        kind: 'MethodImpl',
+        name: nameTok.image,
```
[the remark this answers is not in the document]
Comment (ThePlenkov) [reply packages/aclass/src/parser.ts]: Fixed in cfe5839d — new `readQualifiedName(toks, start)` walks `<name>(~|=>)<name>…` chains and returns the joined `image` plus the last-consumed index. Applied to `parseMethodImpl` AND `parseMethodDecl`. Verified on the petstore3 corpus: the 20 method impls in `zcl_petstore3.clas.abap` now show `constructor` + 19 fully-qualified `zif_petstore3~<op>` names. Live TRL deploy + AUnit still 3/3 pass.
